pkg/azure/aks/pricing_map.go

```diff
@@ -0,0 +1,183 @@
+	logger := parentLogger.With("subsystem", "pricingMap")
+
+	p := &PriceStore{
+		lock:              &sync.RWMutex{},
+		logger:            logger,
+		context:           parentContext,
+		subscriptionId:    subId,
+		retailPriceClient: priceClient,
+
+		RegionMap: make(map[string]PriceByPriority),
+		Cache:     make(map[string]*retailPriceSdk.ResourceSKU),
+	}
+
+	go func() {
+		err := p.PopulatePriceStore([]string{})
+		if err != nil {
+			p.logger.LogAttrs(p.context, slog.LevelError, "error populating initial price store", slog.String("error", err.Error()))
+		}
+	}()
+
+	return p
+}
+
+func (p *PriceStore) buildQueryFilter(locationList []string) string {
+	if len(locationList) == 0 {
+		return `serviceName eq 'Virtual Machines' and priceType eq 'Consumption'`
+	}
+
+	locationListFilter := []string{}
+	for _, region := range locationList {
+		locationListFilter = append(locationListFilter, fmt.Sprintf("armRegionName eq '%s'", region))
+	}
+
+	locationListStr := strings.Join(locationListFilter, " or ")
+	return fmt.Sprintf(`serviceName eq 'Virtual Machines' and priceType eq 'Consumption' and (%s)`, locationListStr)
+}
+
+func (p *PriceStore) buildListOptions(locationList []string) *retailPriceSdk.RetailPricesClientListOptions {
+	queryFilter := p.buildQueryFilter(locationList)
+	return &retailPriceSdk.RetailPricesClientListOptions{
+		APIVersion:  to.StringPtr(AZ_API_VERSION),
+		Filter:      to.StringPtr(queryFilter),
+		MeterRegion: to.StringPtr(`'primary'`),
+	}
+}
+
+func (p *PriceStore) determineMachineOperatingSystem(sku retailPriceSdk.ResourceSKU) MachineOperatingSystem {
+	switch {
+	case strings.Contains(sku.ProductName, "Windows"):
+		return Windows
+	default:
+		return Linux
+	}
+}
+
+func (p *PriceStore) determineMachinePriority(sku retailPriceSdk.ResourceSKU) MachinePriority {
+	switch {
+	case strings.Contains(sku.SkuName, "Spot"):
+		return Spot
+	default:
+		return OnDemand
+	}
+
+}
+
+func (p *PriceStore) PopulatePriceStore(locationList []string) error {
+	startTime := time.Now()
+	p.logger.LogAttrs(p.context, slog.LevelInfo, "populating price map")
+
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	pager := p.retailPriceClient.NewListPager(p.buildListOptions(locationList))
+
+	for pager.More() {
+		page, err := pager.NextPage(p.context)
+		if err != nil {
+			p.logger.LogAttrs(p.context, slog.LevelError, "error paging")
+			return ErrPageAdvanceFailure
+		}
+
+		for _, v := range page.Items {
+			regionName := v.ArmRegionName
+			if regionName == "" {
+				p.logger.LogAttrs(p.context, slog.LevelInfo, "region name for price not found")
+				continue
+			}
+
+			if _, ok := p.RegionMap[regionName]; !ok {
+				p.logger.LogAttrs(p.context, slog.LevelInfo, "populating machine prices for region", slog.String("region", regionName))
+				p.RegionMap[regionName] = make(PriceByPriority)
+				p.RegionMap[regionName][Spot] = make(PriceByOperatingSystem)
+				p.RegionMap[regionName][OnDemand] = make(PriceByOperatingSystem)
```

Comment:
Is there a specific reason to use make and now instantiate an object like `&PriceByPriority`? Trying to think if there's any impact on the ability to update the objects when refreshing the pricing store.

Reply:
I envision this as instantiation because this is a region that isn't present in the RegionMap yet.  It's a little pre-mature because we could have a region without Spot or OnDemand, I suppose?  But we will almost certainly have both in each region.

When refreshing the RegionMap, I envision blowing it away entirely and building it from scratch.  We can think about a more efficient way to do it if we want in the future, but this seems good enough for now?

Reply:
Ah yeah, that's the approach that eks takes. In the `Collect` call, it checks to see if the "store" needs to be refreshed, which then generates a new pricing map. 

I've gotta write some toy code and explore this aspect of Go a bit more. Appreciate the insights